File: python/days/day_01.py
#!/usr/bin/env python3
"""Advent of Code, Day 01"""

from itertools import cycle


def part1():
    with open("../data/d01p1.txt", "r") as f:
        data = list(f)

    return sum([int(x) for x in data])


def part2():
    with open("../data/d01p1.txt", "r") as f:
        data = list(f)

    state = 0
    state_set = {0}

    for item in cycle(data):
        state += int(item)
        if state in state_set:
            return state
        state_set.add(state)


if __name__ == "__main__":
    print("part 1:", part1())
    print("part 2:", part2())

File: rust/advent-of-code/src/day01.rs
// Day 01

use std::collections::HashSet;

pub fn part1() {
    let input = include_str!("data/d01p1.txt");
    let x: i32 = input.lines()
        .map(|i: &str| i.parse::<i32>().unwrap())
        .sum();

    println!("part 1: {}", x);
}

pub fn part2() {
    let input = include_str!("data/d01p1.txt");
    let endless = input.lines()
        .map(|i: &str| i.parse::<i32>().unwrap())
        .cycle();

    let mut state = 0;
    let mut states = HashSet::new();
    for item in endless {
        state += item;
        if ! states.insert(state) {
            println!("part 2: {}", state);
            return
        }
    }
}